File: Algorithms_python/46_Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res = []
        path = [0] * n
        used = [False] * n
        def dfs(u):
            if u == n:
                res.append(path.copy())
                return
            for i in range(n):
                if not used[i]:
                    path[u] = nums[i]
                    used[i] = True
                    dfs(u + 1)
                    used[i] = False
                    
        dfs(0)
        return res
'''
dfs深度搜索算法，一层层往下枚举
我们设计一个函数dfs(i)表示已经填完了前i个位置，现在需要填第i+1个位置。
枚举所有可能的数，如果这个数没有被填过，就填入这个数，然后继续填下一个位置，直到填完所有的位置。
'''

File: Algorithms_C#/46_Permutations.cs
public class Solution {
    IList<IList<int>> permutations = new List<IList<int>>();
    IList<int> temp = new List<int>();
    int n;

    public IList<IList<int>> Permute(int[] nums) {
        foreach (int num in nums){
            temp.Add(num);
        }
        n = nums.Length;
        dfs(0);
        return permutations;
    }

    public void dfs(int index){
        if (index == n){
            permutations.Add(new List<int>(temp));
        }
        else{
            for (int i = index; i < n; ++i){
                Swap(temp, index, i);
                dfs(index + 1);
                Swap(temp, index, i);
            }
        }
    }

    public void Swap(IList<int> temp, int index1, int index2) {
        int curr = temp[index1];
        temp[index1] = temp[index2];
        temp[index2] = curr;
    }
}